### DISCUSSION TITLE
Can a controller view render into multiple zones?

### DISCUSSION THREAD

**Discussion by kevinsmgov** [OrchardCMS/OrchardCore#9373] [Q&A]
I have a site with a complex theme and placement.json file. Different content types render their parts and fields into several different zones in the layout.

I am trying to create a controller to produce custom results based on URL query parameters and I need the results to behave like my existing content types and render with my existing razor templates, following the rules in my existing placement.json file.

To some extent, this works. The controller's resulting viewmodel contains a number of different shapes that present themselves as known parts and fields and those parts and fields render with their expected razor templates. However, it seems like the placement.json file is ignored in the case of the controller's results and everything renders within the Content zone.

It may be that there are additional properties I need to set with the shapefactory in the controller relative to each of the content parts and fields I am returning but I haven't been able to determine what additional properties in the shapes are necessary.

**Comment by ns8482e** [comment]
Controller's `View` is rendered in `Content` zone. However any shape rendered within using `@await DisplayAsync(Model.SomeShape)` will rendered in zone based on placement rule.

To target the Layout zone prefix zone with `/` e.g. `"place" : "/Header:1"`

**Reply by kevinsmgov** [reply]
I guess I didn't specify in my original post that my controller's view razor page is using a couple`@await DisplayAsync(Model.SomeShape)` expressions for different shapes. They render with the correct templates but they are not rendering in the correct zones.

**Reply by ns8482e** [reply]
Without looking at your code it's hard to tell. 

However I feel like you are not using `DisplayDriver` and `DisplayManager`.  To apply placements to your shapes you need to use `DisplayManager` and build display using  `BuildDisplayAsync` to create your shapes.

**Reply by kevinsmgov** [reply]
Based on some other conversations, I was coming to that conclusion as well. However, I haven't been able to find any good examples of controllers using `DisplayManager` and including other content parts. For example, I'm trying to return a collection of content items (`IEnumerable<dynamic>`) and associated Pager, also a `TitlePart `, and a `LinkField `(the `LinkField` needs to go in a different zone).

I can use the `ShapeFactory `to assemble these together into a ViewModel with dynamic properties (this renders but without placement.json and correct zones) but when I try and pass that ViewModel to `BuildDisplayAsync`, I only get a result with the metadata, no payload.

**Reply by ns8482e** [reply]
If you are using `ShapeFactory`, you can still render `shape` to specific `zone` using  `<zone>` tag helper from view

In controller
```cs
model.SomeShape = shapeFactory.SomeShape();
model.AnotherShape = shapeFactory.AnotherShape();
model.ContentShape = shapeFactory.ContentShape();

return View(model);
```
In razor

```razor
<zone name="Footer"> 
<!-- render in footer -->
     @await DisplayAsync(Model.SomeShape)
</zone>
<h1>This is rendered in content zone</h1>

<zone name="Header"> 
<!-- render in header-->
     @await DisplayAsync(Model.AnotherShape)
</zone>

<div>this is content</div>
@await DisplayAsync(Model.ContentShape)
```

**Reply by kevinsmgov** [reply]
True; and the controller can also inject shapes directly:

```
var layout = await _layoutAccessor.GetLayoutAsync();
await layout.Zones["Featured"].AddAsync(Model.SomeShape)
await layout.Zones["Related"].AddAsync(Model.SomeOtherShape)
```

However, if possible, I'd like my placement.json file to be the master arbitrator of templates and zones for all my content, even if it's being displayed via a parameterized URL and a controller. That way, when our communications group comes in with a new design that relocates different content types in the layout, I'm not hunting through a bunch of source code for zone assignments.

I'm working under the assumption that if I can make my controllers/drivers function like `OrchardCore.Contents` then, theoretically, their results should also honor placement.json. 

Of course, it might not be possible. It's good to know that I have fallbacks.

It's kind of ironic but, when I first started using Orchard, the placement.info file was a difficult concept to grasp and a bit frustrating. It was much easier to just drive things inline and render pages as big monolithic templates. Now though, I really appreciate it's ability to manage a design and assemble pages like so many Lego blocks. 

It's similar, I guess, to a novice HTML designer moving from inline styles to CSS. There's a learning curve and frustration at the beginning but once you get it down it's easier overall to manage your designs with a central CSS file.

**Reply by ns8482e** [reply]
> Of course, it might not be possible. It's good to know that I have fallbacks.

It is possible and `OrchardCore` has that feature built-in. However you need to walk little extra mile.

- If `TModel` in your controller is `ContentItem` then you can inject `IContentItemDisplayManager` and use `BuildDisplayAsync` to create your Detail/Summary shapes for your content item.

- If `TModel` in your controller is `POCO` then you can inject `IDisplayManager<T>` and use `BuildDisplayAsync` to create your shapes for `POCO` model

For  `POCO` you need to define your own Drivers and register them in Startup. 

e.g. If you have `ProductModel` c# class you can define different drivers that targets different shapes as below in startup

```cs
services.AddScoped<IDisplayManager<ProductModel>, DisplayManager<ProductModel>>();
services.AddScoped<IDisplayDriver<ProductModel>, ProductDetailDisplayDriver>();
services.AddScoped<IDisplayDriver<ProductModel>, ProductSpecDisplayDriver>();
services.AddScoped<IDisplayDriver<ProductModel>, ProductMediaDisplayDriver>();
  ```

and in Controller,

```cs
// create product instance from db or query string
var product = new ProductModel();
//Build shape using injected display manager
var prodcutShape = productDM.BuildDisplayAsync(product, null, "Detail" );
// return shape to view
return View(prodcutShape);
```
In `DisplayDriver` you can define `ShapeType` , `Alternates`, `Wrappers` and default `Location` that you can override using placement.json or Admin placement feature.

**Reply by kevinsmgov** [reply]
I think that is just what I'm looking for. I'll prototype and see.

**Reply by ns8482e** [reply]
@kevinsmgov I saw you notification, I don't know why I am not seeing your message here!  

In notification, I see that you were using core shape "List" for your VM in driver - I would suggest to define your own shapes from your Driver instead of using core shape. It will resolve the issue.

**Reply by kevinsmgov** [reply]
When I noticed that myself, I deleted the message. However, now my driver is returning a shape with a Content property but still no other properties from the viewmodel. Am I mistaken in thinking I can use `Initialize` in my driver to return a strongly typed shape?

**Reply by ns8482e** [reply]
Correct! your VM shape is in Content Zone. The `return View(shape)`  from Controller returns view with `ZoneHolding` model and Your shape/vm is placed in `Content` zone as defined in the `Driver`.

From controller view call `@await DisplayAsync(Model)` that will call your shape defined in driver.

i.e. In driver if line `return Initialize<TModel>("QueryListView", vm=> {  }) `  defines shape type `QueryListView` then Create View`QueryListView.cshtml` or  `QueryListView.liquid` and your model will be available in that view.

**Reply by kevinsmgov** [reply]
I have all those things and the template is showing the Content shape in the model but none of the other ViewModel properties are coming along with it.

Would you be interested in engaging in a quick consulting gig on this? If so, email me.

**Reply by ns8482e** [reply]
Sure, I can help you with

**Reply by kevinsmgov** [reply]
Coming back to the original question about rendering in multiple zones, does `IDisplayManager` interact with _placement.json_ the way that `IContentItemDisplayManager` does? If so, where does that process occur in the pipeline?

**Reply by ns8482e** [reply]
@kevinsmgov  Yes IMHO both should have same functionality, as `Placement` and `DisplayManager<T>` is part of OC's `DisplayManagement`, and `IContentItemDisplayManager` is built on top of  OC's `DisplayManagement`.

See, how `DisplayManager<T>` and `ContentItemDisplayManager` both inherited from same base class `BaseDisplayManager`.

https://github.com/OrchardCMS/OrchardCore/blob/a779df8beb42977efbb053a12e6fd7daac92810e/src/OrchardCore/OrchardCore.DisplayManagement/DisplayManager.cs#L12

https://github.com/OrchardCMS/OrchardCore/blob/a779df8beb42977efbb053a12e6fd7daac92810e/src/OrchardCore/OrchardCore.ContentManagement.Display/ContentItemDisplayManager.cs#L25

And binds to  Placement

https://github.com/OrchardCMS/OrchardCore/blob/a779df8beb42977efbb053a12e6fd7daac92810e/src/OrchardCore/OrchardCore.DisplayManagement/DisplayManager.cs#L59

https://github.com/OrchardCMS/OrchardCore/blob/a779df8beb42977efbb053a12e6fd7daac92810e/src/OrchardCore/OrchardCore.ContentManagement.Display/ContentItemDisplayManager.cs#L90

**Reply by kevinsmgov** [reply]
OK, here's the tricky part. In addition to the "FilteredNewsList" POCO object (an `IEnumerable<ContentItem>` collection and a `Pager` object) I need to return a `LinkField` and a `TitlePart` that need to go into different zones.

Assumably, I'll need a BuildDisplayAsync call for all three of these objects. In the Views folder of my module, I have the Index.cshtml
```
@model NewsControllerResult
@await DisplayAsync(Model.ShapeLinkField)
@await DisplayAsync(Model.ShapeTitlePart)
@await DisplayAsync(Model.ShapeFilteredNewsList)
```
I tried creating a driver for the LinkField but I can't segregate it to just this operation. So, I'll want to use the built in LinkField and TitlePart drivers. How can I front-load the data I pass to `BuildDisplayAsync` so that it picks up the right field name and differentiator?

**Reply by ns8482e** [reply]
@kevinsmgov You can create a `ContentItem` with `LinkField` and `TitlePart`. Query the item by Id in your controller and  build display for it, so your POCO `ViewModel` can have three fields.

- ShapeFilteredNewsList 
- ContentItemShape 
- PagerShape

**Reply by kevinsmgov** [reply]
I was holding that approach in reserve, hoping there was a way to generate all this programmatically (or, at least understand the system enough to do that).

Probably the best approach then would be to create "Filtered Content Header" item for the different filter controllers and manage them as Admin settings.

Thanks for all the information.

**Reply by ns8482e** [reply]
Refer to `AdminController` in Flow module - that creates content item dynamically (not stored in db) and builds display and returns html as ajax response.

**Reply by kevinasdzine** [reply]
I got a working solution by returning a collection of IShape from my controller and then simply running this in the controller's view

```
@foreach (var zone in Model)
{
    @await DisplayAsync(zone)
}
```

And that covered getting shapes to global zones fine (e.g. "/Header"). However, I have a use-case now where I want to manage local zones generically (as they are from Orchard Core ContentTypes) so I can create a similar structure (e.g. article element contains header and content and footer where shapes position in their relevant local zones).

When I look at that AdminController you suggested, it's creating a ContentCard but I'm looking for some example code where a controller is creating something similar to a native ContentType shape (is that what the ItemController in OrchardCore.Contents is creating?).
